In [ ]:
%%bash
pwd

#Present working directory

/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell


In [ ]:
%%bash
ls

#List files in current folder

Linux_shell_coursew1.ipynb
SC_Week0_Unit0.ipynb
SC_Week0_Unit1.ipynb
SC_Week0_Unit2.ipynb
SC_Week0_Unit4.ipynb
SC_Week0_Unit5.ipynb


In [ ]:
%%bash
clear # used to clear the screen
# or use ctrl+L

In [ ]:
%%bash
exit
# or ctrl+D to come out of shell

In [ ]:
%%bash
ls -a
# -a option gives hidden files also

.
..
Linux_shell_course.ipynb
SC_Week0_Unit0.ipynb
SC_Week0_Unit1.ipynb
SC_Week0_Unit2.ipynb
SC_Week0_Unit4.ipynb
SC_Week0_Unit5.ipynb


In [ ]:
%%bash
ls -l
# -l option gives long format

total 270
-rw-r--r-- 0 kavis 197609  11069 Oct  6 11:21 Linux_shell_course.ipynb
-rw-r--r-- 0 kavis 197609   8173 Oct  5 19:51 SC_Week0_Unit0.ipynb
-rw-r--r-- 0 kavis 197609 230730 Oct  5 19:51 SC_Week0_Unit1.ipynb
-rw-r--r-- 0 kavis 197609  16186 Oct  5 20:28 SC_Week0_Unit2.ipynb
-rw-r--r-- 0 kavis 197609   3531 Oct  5 19:52 SC_Week0_Unit4.ipynb
-rw-r--r-- 0 kavis 197609   4159 Oct  5 19:51 SC_Week0_Unit5.ipynb


In [ ]:
%%bash
ls --help
#summary of how to use the command

Usage: ls [OPTION]... [FILE]...
List information about the FILEs (the current directory by default).
Sort entries alphabetically if none of -cftuvSUX nor --sort is specified.

Mandatory arguments to long options are mandatory for short options too.
  -a, --all                  do not ignore entries starting with .
  -A, --almost-all           do not list implied . and ..
      --author               with -l, print the author of each file
  -b, --escape               print C-style escapes for nongraphic characters
      --block-size=SIZE      with -l, scale sizes by SIZE when printing them;
                               e.g., '--block-size=M'; see SIZE format below
  -B, --ignore-backups       do not list implied entries ending with ~
  -c                         with -lt: sort by, and show, ctime (time of last
                               modification of file status information);
                               with -l: show ctime and sort by name;
                               othe

In [ ]:
%%bash
man ls
clear
# doesn't work with vs code
# manual page for the command, which provides detailed documentation.

LS(1)                     User Commands                    LS(1)

NAME
       ls - list directory contents

SYNOPSIS
       ls [OPTION]... [FILE]...

DESCRIPTION
       List  information  about the FILEs (the current directory
       by default).  Sort entries alphabetically if none of -cf‐
       tuvSUX nor --sort is specified.

       Mandatory arguments to long  options  are  mandatory  for
       short options too.

       -a, --all
              do not ignore entries starting with .

       -A, --almost-all
              do not list implied . and ..

       --author
              with -l, print the author of each file

       -b, --escape
              print C-style escapes for nongraphic characters

       --block-size=SIZE
              with  -l,  scale sizes by SIZE when printing them;
              e.g., '--block-size=M'; see SIZE format below

       -B, --ignore-backups
              do not list implied entries ending with ~

       -c     with -lt: sort by, and show, ctime 

Filesystem Hierarchy Standard followed by all linux systems
FHS 3.0 released on June 03, 2015
Available at
https://refspecs.linuxfoundation.org/fhs.shtml

In [ ]:
%%bash
cd ..
##change directory to parent directory of current directory (parent directory means just one level up)
pwd

cd . 
## single dot stands for present directory so, no change
# cd . means stay in the current directory.
pwd


cd mnt/c/Users/kavis/Upskilling/'Github repository'/Rep1
pwd
#this path is relative to current directory and it doesn't actually exist

cd /mnt/c/Users/kavis/Upskilling/'Github repository'/Rep1
pwd
#absolute path

# use the command cd ~ or cd $HOME
cd ~
# The home directory is the starting directory for a user
pwd


/mnt/c/Users/kavis/Upskilling/Github repository/Rep1
/mnt/c/Users/kavis/Upskilling/Github repository/Rep1


bash: line 11: cd: mnt/c/Users/kavis/Upskilling/Github repository/Rep1: No such file or directory


/mnt/c/Users/kavis/Upskilling/Github repository/Rep1
/mnt/c/Users/kavis/Upskilling/Github repository/Rep1
/home/kavis


In [17]:
%%bash
pwd

cd     # cd without arguement takes to the home directory
pwd

cd /    #root directory  (/) is the highest level in the filesystem hierarchy 
pwd
ls -a

#When you list all items in a directory using ls -a, you will see ..:
#   . represents the current directory.
#  .. always represents the parent directory, even if it’s a reference to the same directory (as is the case in the root directory).

cd /
pwd
cd ..
pwd

/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
/home/kavis
/
.
..
bin
bin.usr-is-merged
boot
dev
etc
home
init
lib
lib.usr-is-merged
lib64
lost+found
media
mnt
opt
proc
root
run
sbin
sbin.usr-is-merged
snap
srv
sys
tmp
usr
var
/
/


In [ ]:
%%bash
pwd
cd /c
ls -l

/g/My Drive/Upskilling/Github repository/Rep1/Linux_shell
total 10793673
drwxr-xr-x 1 kavis 197609          0 Oct 20  2022 $Recycle.Bin
-rw-r--r-- 1 kavis 197609         40 Oct 21  2022 2CA0AE717CAA
lrwxrwxrwx 1 kavis 197609          8 Oct 11  2022 Documents and Settings -> /c/Users
drwxr-xr-x 1 kavis 197609          0 Jul 30  2022 Drivers
-rw-r--r-- 1 kavis 197609      12288 Jul 12 09:14 DumpStack.log
-rw-r--r-- 1 kavis 197609      12288 Oct  5 18:18 DumpStack.log.tmp
drwxr-xr-x 1 kavis 197609          0 May  4 13:43 OneDriveTemp
drwxr-xr-x 1 kavis 197609          0 May  7  2022 PerfLogs
drwxr-xr-x 1 kavis 197609          0 Oct  6 09:13 Program Files
drwxr-xr-x 1 kavis 197609          0 Jun 11 08:47 Program Files (x86)
drwxr-xr-x 1 kavis 197609          0 Sep 19 22:02 ProgramData
drwxr-xr-x 1 kavis 197609          0 Aug 10  2023 Recovery
-rw-r--r-- 1 kavis 197609       1505 Dec 19  2023 Setup.log
drwxr-xr-x 1 kavis 197609          0 Oct  6 09:12 System Volume Information
drwxr-xr-x 1 

#### Structure of Each Line:
1. **Permissions**: (`drwxr-xr-x`)
   - This shows the file permissions for the owner, group, and others.
   - first letter -  `d` indicates a **directory** (if it were a file, it would be `-`).
                     'l' indicates symbolic link (shortcut)
                     '-' indicates file 
                     'c' character device (terminal)
                     'b' block devices (hard discs)
   - next three for owner - `rwx` means **read, write, and execute** permissions for the owner.
   - next three for group - `r-x` means **read and execute** permissions for the group.
   - next three for others- `r-x` means **read and execute** permissions for others.

   rwx (7)
   rw- (6)
   r-x (5)
   r-- (4)
   --x (1)

2. **Links**: (`1`)
   - The number of links or directories inside this directory. For a directory, this number (1) represents how many subdirectories, parent directories, or other references link to it. A hard link is like an alias or shortcut that points to the same file or directory in the filesystem. 

3. **Owner**: (`kavis`)
   - The username of the owner of the file or directory.

4. **Group**: (`197609`)
   - The group ID associated with the file.

5. **Size**: (`0`, `40`, `12288`, etc.)
   - The size of the file in bytes. For directories, it’s usually `0` since it doesn’t directly reflect the size of contents.

6. **Timestamp**: (`Oct 6 14:21`)
   - The date and time when the file or directory was last modified.

7. **File/Directory Name**: (`$Recycle.Bin`, `Documents and Settings`, etc.)
   - The name of the file or directory.

#### Listing Explanation:

**`Documents and Settings -> /c/Users`**:
   - This is a **symbolic link** (shortcut) that points to `/c/Users`. It means that trying to access `Documents and Settings` will actually redirect you to the `/c/Users` directory.

**`Drivers`**:
   - Likely a folder containing drivers for the system. Drivers are software that help the operating system communicate with hardware devices.

**`Program Files`** and **`Program Files (x86)`**:
   - Standard directories in Windows for installing 64-bit and 32-bit applications, respectively.

**`ProgramData`**:
   - A directory used to store application data that is not specific to a user. It’s shared among all users of the system.

**`Users`**:
    - This is where user profiles are stored on Windows, such as `Desktop`, `Documents`, `Downloads`, and so on for each user.

**`Windows`**:
    - The main directory containing the Windows operating system files.

**`hiberfil.sys`**:
    - A **system file** used by Windows to store the system state when it goes into **hibernation**. This file is often large and contains everything stored in memory (RAM).

In [4]:
%%bash
pwd
cd /
ls -l
ls -i  #ls -i shows the inode number associated with each file or directory, which helps identify them uniquely
cd -  # go to previous directory

/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
total 2284
lrwxrwxrwx   1 root root       7 Apr 22 13:08 bin -> usr/bin
drwxr-xr-x   2 root root    4096 Feb 26  2024 bin.usr-is-merged
drwxr-xr-x   2 root root    4096 Apr 22 13:08 boot
drwxr-xr-x  16 root root    3560 Oct  6 11:37 dev
drwxr-xr-x  92 root root    4096 Oct  6 16:28 etc
drwxr-xr-x   3 root root    4096 Oct  5 12:54 home
-rwxrwxrwx   1 root root 2260248 Sep 27 00:55 init
lrwxrwxrwx   1 root root       7 Apr 22 13:08 lib -> usr/lib
drwxr-xr-x   2 root root    4096 Apr  8 14:37 lib.usr-is-merged
lrwxrwxrwx   1 root root       9 Apr 22 13:08 lib64 -> usr/lib64
drwx------   2 root root   16384 Oct  5 12:49 lost+found
drwxr-xr-x   2 root root    4096 Sep 27 20:13 media
drwxr-xr-x   5 root root    4096 Oct  5 12:49 mnt
drwxr-xr-x   2 root root    4096 Sep 27 20:13 opt
dr-xr-xr-x 313 root root       0 Oct  6 11:37 proc
drwx------   4 root root    4096 Oct  5 12:49 root
drwxr-xr-x  18 root root     580 Oct  6 12:23

# inode and hardlink
The inode tracks the file or directory, while the hard link count tells you how many different references or paths point to that inode.

## inode
Each inode number (e.g., 12, 8193, 8194) uniquely identifies a file or directory on the filesystem, like a unique ID.

ls -i shows the inode number associated with each file or directory, which helps identify them uniquely


## hard link
Each directory starts with at least 2 hard links:

One for the directory itself (.)
One for its parent directory (..)
Every subdirectory inside it adds 1 more hard link to the count. This is because each subdirectory contains a link (..) pointing back to the parent directory. 

Example:
var has 13 hard links.
It includes . and ...
And 11 subdirectories inside var, each adding 1 to the count.

When you use ls -l, the number of hard links is shown in the second column. For example, in drwxr-xr-x 13 root root 4096 Oct 5 12:49 var, the 13 means there are 13 hard links to the var directory. The inode for var is linked to every file or directory that shares it. 

/bin -       essential command binaries - files in machine language for processor
/usr/bin -   User commands
/usr/lib -   Libraries

In [2]:
%%bash
date
date -R

cal

cal sep 1991

ncal sept 2024


Sun Oct  6 14:36:14 UTC 2024
Sun, 06 Oct 2024 14:36:14 +0000
    October 2024      
Su Mo Tu We Th Fr Sa  
       1  2  3  4  5  
 6  7  8  9 10 11 12  
13 14 15 16 17 18 19  
20 21 22 23 24 25 26  
27 28 29 30 31        
                      
   September 1991     
Su Mo Tu We Th Fr Sa  
 1  2  3  4  5  6  7  
 8  9 10 11 12 13 14  
15 16 17 18 19 20 21  
22 23 24 25 26 27 28  
29 30                 
                      
    September 2024    
Su  1  8 15 22 29   
Mo  2  9 16 23 30   
Tu  3 10 17 24      
We  4 11 18 25      
Th  5 12 19 26      
Fr  6 13 20 27      
Sa  7 14 21 28      
               total        used        free      shared  buff/cache   available
Mem:         7993720     1832652     6127780        3104      303492     6161068
Swap:        2097152           0     2097152
kavis adm dialout cdrom floppy sudo audio dip video plugdev users netdev


In [3]:
%%bash

free
free -h
groups

               total        used        free      shared  buff/cache   available
Mem:         7993720     1705600     6226108        3156      332388     6288120
Swap:        2097152           0     2097152
               total        used        free      shared  buff/cache   available
Mem:           7.6Gi       1.6Gi       5.9Gi       3.1Mi       324Mi       6.0Gi
Swap:          2.0Gi          0B       2.0Gi
kavis adm dialout cdrom floppy sudo audio dip video plugdev users netdev


In [12]:
%%bash
mkdir level1
ls -l
# permission string reflects permissions to users which can be changed my chmod
chmod g-w level1
ls -l


mkdir: cannot create directory ‘level1’: File exists


total 324
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis  35547 Oct  6 14:32 Linux_shell_w1.ipynb
-rwxrwxrwx 1 kavis kavis   8173 Oct  5 14:21 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis  16186 Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavis kavis   3531 Oct  5 14:22 SC_Week0_Unit4.ipynb
-rwxrwxrwx 1 kavis kavis   4159 Oct  5 14:21 SC_Week0_Unit5.ipynb
drwxrwxrwx 1 kavis kavis   4096 Oct  6 17:37 level1
total 324
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis  35547 Oct  6 14:32 Linux_shell_w1.ipynb
-rwxrwxrwx 1 kavis kavis   8173 Oct  5 14:21 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis  16186 Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavis kavis   3531 Oct  5 14:22 SC_Week0_Unit4.ipynb
-rwxrwxrwx 1 kavis kavis   4159 Oct  5 14:21 SC_

# Difference Between Linux and Windows File Systems

### File System Types:

Linux: Uses file systems like ext4, ext3, xfs, and btrfs.
Windows: Uses NTFS (New Technology File System) and FAT32.


### Permissions
Linux: Has a detailed permission model with read, write, and execute permissions for the user, group, and others. It supports commands like chmod and chown to control access.
Windows: Uses Access Control Lists (ACLs) to set permissions. ACLs allow for more granular control but are not compatible with the Unix permission model. This is why changing permissions with chmod on a Windows file system has no effect.


### Case Sensitivity:

Linux: Case-sensitive. file.txt and File.txt are treated as separate files.
Windows: Generally case-insensitive. file.txt and File.txt are considered the same file.

### File Path Differences:

Linux: Paths use forward slashes (/), e.g., /home/kavis.
Windows: Paths use backslashes (\), e.g., C:\Users\kavis.

When you work in directories like /mnt/c/, /mnt/d/, etc., you’re actually interacting with the Windows file system through WSL, which has limitations when it comes to Unix-style operations.
By working in Linux directories like /home/kavis, you can use native Linux commands and features (like chmod) without restrictions.
In Summary: Move to directories like /home/kavis to use Linux file system features effectively and see your chmod changes reflect properly!

### Move to linux file system
To move to the Linux file system in WSL, you can navigate to the home directory or any other directory that resides within the Linux environment. Here's how you can do it:

Go to Your Linux Home Directory: cd~
Create a New Directory: Once you're in your Linux home directory, create a new directory for testing:


In [5]:
%%bash

cd ~
pwd

cd -    # change to previous directory
pwd

cd 
pwd

rm -r level1
mkdir level1
ls -l

chmod g-x level1
ls -l

chmod g-r level1
ls -l

chmod o+w level1
ls -l

chmod 700 level1
ls -l

chmod 707 level1
ls -l


/home/kavis
/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
/home/kavis
total 4
drwxr-xr-x 2 kavis kavis 4096 Oct  6 19:09 level1
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
total 4
drwxr--r-x 2 kavis kavis 4096 Oct  6 19:09 level1
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
total 4
drwx---r-x 2 kavis kavis 4096 Oct  6 19:09 level1
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
total 4
drwx---rwx 2 kavis kavis 4096 Oct  6 19:09 level1
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
total 4
drwx------ 2 kavis kavis 4096 Oct  6 19:09 level1
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
total 4
drwx---rwx 2 kavis kavis 409

In [11]:
%%bash
cd 
pwd

cd level1 
pwd
ls -l

touch file1 #make empty file - touch is actually used to chage timestamp
chmod 707 file1
pwd
ls -l

cp file1 file2
pwd
ls -l

mv file2 ..
pwd
ls -l

cd ..
pwd
ls -l

mv file2 file2a
ls -l

cd level1 
rm file1
cd ..
rm file2a

/home/kavis
/home/kavis/level1
total 0
/home/kavis/level1
total 0
-rwx---rwx 1 kavis kavis 0 Oct  8 02:37 file1
/home/kavis/level1
total 0
-rwx---rwx 1 kavis kavis 0 Oct  8 02:37 file1
-rwx---r-x 1 kavis kavis 0 Oct  8 02:37 file2
/home/kavis/level1
total 0
-rwx---rwx 1 kavis kavis 0 Oct  8 02:37 file1
/home/kavis
total 4
-rwx---r-x 1 kavis kavis    0 Oct  8 02:37 file2
drwx---rwx 2 kavis kavis 4096 Oct  8 02:37 level1
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
total 4
-rwx---r-x 1 kavis kavis    0 Oct  8 02:37 file2a
drwx---rwx 2 kavis kavis 4096 Oct  8 02:37 level1
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1


In [2]:
%%bash
cd 
pwd
touch "file 2"  #space in filename
ls -li

alias rm="rm -i"
rm "file 2"
alias
pwd
ls -lia

mkdir level1a
ls -lia
# as new directory is added, hard links for present directory . increased by one


rm -r level1a

/home/kavis
total 4
43698 -rw-r--r-- 1 kavis kavis    0 Oct  6 19:05 file 2
43662 drwx---rwx 2 kavis kavis 4096 Oct  6 18:41 level1
43664 lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
alias rm='rm -i'
/home/kavis
total 56
  686 drwxr-x--- 10 kavis kavis 4096 Oct  6 19:05 .
24577 drwxr-xr-x  3 root  root  4096 Oct  5 12:54 ..
46407 -rw-------  1 kavis kavis 1484 Oct  6 19:03 .bash_history
46419 -rw-r--r--  1 kavis kavis  220 Oct  5 12:54 .bash_logout
 1227 -rw-r--r--  1 kavis kavis 3771 Oct  5 12:54 .bashrc
46417 drwx------  3 kavis kavis 4096 Oct  5 13:50 .cache
82512 drwxr-xr-x  3 kavis kavis 4096 Oct  6 12:03 .dotnet
82473 drwxr-xr-x  3 kavis kavis 4096 Oct  6 12:06 .ipython
46428 drwxr-xr-x  2 kavis kavis 4096 Oct  6 10:49 .landscape
31011 drwxr-xr-x  3 kavis kavis 4096 Oct  6 11:12 .local
46423 -rw-r--r--  1 kavis kavis    0 Oct  6 10:49 .motd_shown
46422 -rw-r--r--  1 kavis kavis  807 Oct  5 12:54 .profile
46410 -rw-r--r-- 

In [16]:
%%bash
whoami
cd /   # move to root directory
pwd
ls -la
# root folder doesnt have rwx permissions for kavis- it has permission only for root directory . 
touch file1
# write w not allowed in root directory . by other users 


kavis
/
total 2292
drwxr-xr-x  22 root root    4096 Oct  6 11:37 .
drwxr-xr-x  22 root root    4096 Oct  6 11:37 ..
lrwxrwxrwx   1 root root       7 Apr 22 13:08 bin -> usr/bin
drwxr-xr-x   2 root root    4096 Feb 26  2024 bin.usr-is-merged
drwxr-xr-x   2 root root    4096 Apr 22 13:08 boot
drwxr-xr-x  16 root root    3560 Oct  6 11:37 dev
drwxr-xr-x  92 root root    4096 Oct  6 19:03 etc
drwxr-xr-x   3 root root    4096 Oct  5 12:54 home
-rwxrwxrwx   1 root root 2260248 Sep 27 00:55 init
lrwxrwxrwx   1 root root       7 Apr 22 13:08 lib -> usr/lib
drwxr-xr-x   2 root root    4096 Apr  8 14:37 lib.usr-is-merged
lrwxrwxrwx   1 root root       9 Apr 22 13:08 lib64 -> usr/lib64
drwx------   2 root root   16384 Oct  5 12:49 lost+found
drwxr-xr-x   2 root root    4096 Sep 27 20:13 media
drwxr-xr-x   5 root root    4096 Oct  5 12:49 mnt
drwxr-xr-x   2 root root    4096 Sep 27 20:13 opt
dr-xr-xr-x 313 root root       0 Oct  6 11:37 proc
drwx------   4 root root    4096 Oct  5 12:49 root
drwxr

touch: cannot touch 'file1': Permission denied


CalledProcessError: Command 'b'whoami\ncd /   # move to root directory\npwd\nls -la\n# root folder doesnt have rwx permissions for kavis- it has permission only for root directory\ntouch file1\n# write not allowed in root \n'' returned non-zero exit status 1.

In [13]:
%%bash

cd 
pwd
# Create a test directory and navigate into it
mkdir -p test_dir/level1
cd test_dir
pwd

# Create some files and subdirectories
touch file1 file2 file3
mkdir subdir1 subdir2
ls # Default listing of files and directories in the current directory

# List in long format (-l) showing permissions, owner, group, size, and timestamp
ls -l

# List with human-readable file sizes (-lh) to display sizes in KB, MB, etc.
ls -lh

# Show hidden files as well (-a), includes files starting with '.'
ls -a

# Long format with hidden files (-la), useful to see permissions of hidden files
ls -la

# List in long format with hidden files and include inode number (-lia)
ls -lia

# Show the inode numbers (-i) of files and directories in the current directory
ls -i

# List in long format of only directories, showing details of directories themselves (-ld)
ls -ld */

# Show inode numbers along with hidden files in a directory listing (-ia)
ls -ia

# Show inode numbers along with detailed view for hidden files (-lia)
ls -lia

# Show file sizes in blocks (-s)
ls -s

# Recursive listing, show contents of all directories and subdirectories (-R)
ls -R

# Recursive listing in long format (-lR)
ls -lR

# Show files sorted by size (-lS) to get largest files first
ls -lS

# List files sorted by modification time, newest first (-lt)
ls -lt

# List files in reverse order by modification time (-ltr)
ls -ltr

# Combine -i and -l to show inode numbers along with detailed view (-il)
ls -il

# Combine -i, -a, and -l to show inode numbers, hidden files, and detailed view (-ial)
ls -ial

# Using a directory as an argument, list contents of "level1" directory
ls level1

# Navigate into the "level1" directory and list contents
cd level1
pwd
ls -l

# Go back to parent directory and list contents recursively with details (-lR)
cd ..
ls -lR

# Use -d to show information only about the directory itself, not its contents
ls -ld level1

# List all files including hidden ones and inodes, but just at the top-level directory (-iAd)
ls -iAd *

# List files with detailed view and human-readable sizes, sort by size (-lhS)
ls -lhS

# List directories themselves in detailed view with inode numbers (-ild)
ls -ild */

# Clean up: remove the created test directory
cd ..
rm -r test_dir


/home/kavis
/home/kavis/test_dir
file1
file2
file3
level1
subdir1
subdir2
total 12
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:46 file1
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:46 file2
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:46 file3
drwxr-xr-x 2 kavis kavis 4096 Oct  8 02:46 level1
drwxr-xr-x 2 kavis kavis 4096 Oct  8 02:46 subdir1
drwxr-xr-x 2 kavis kavis 4096 Oct  8 02:46 subdir2
total 12K
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:46 file1
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:46 file2
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:46 file3
drwxr-xr-x 2 kavis kavis 4.0K Oct  8 02:46 level1
drwxr-xr-x 2 kavis kavis 4.0K Oct  8 02:46 subdir1
drwxr-xr-x 2 kavis kavis 4.0K Oct  8 02:46 subdir2
.
..
file1
file2
file3
level1
subdir1
subdir2
total 20
drwxr-xr-x  5 kavis kavis 4096 Oct  8 02:46 .
drwxr-x--- 11 kavis kavis 4096 Oct  8 02:46 ..
-rw-r--r--  1 kavis kavis    0 Oct  8 02:46 file1
-rw-r--r--  1 kavis kavis    0 Oct  8 02:46 file2
-rw-r--r--  1 kavis kavis    0 Oct  8 02:46 file3
drwxr-xr-x 

In [12]:
%%bash
cd /
cd var/log
pwd 
less alternatives.log  # to read a file

/var/log
update-alternatives 2024-09-27 20:13:01: run with --install /usr/bin/awk awk /usr/bin/mawk 5 --slave /usr/share/man/man1/awk.1.gz awk.1.gz /usr/share/man/man1/mawk.1.gz --slave /usr/bin/nawk nawk /usr/bin/mawk --slave /usr/share/man/man1/nawk.1.gz nawk.1.gz /usr/share/man/man1/mawk.1.gz
update-alternatives 2024-09-27 20:13:01: link group awk updated to point to /usr/bin/mawk
update-alternatives 2024-09-27 20:13:05: run with --install /usr/bin/which which /usr/bin/which.debianutils 0 --slave /usr/share/man/man1/which.1.gz which.1.gz /usr/share/man/man1/which.debianutils.1.gz --slave /usr/share/man/de/man1/which.1.gz which.de1.gz /usr/share/man/de/man1/which.debianutils.1.gz --slave /usr/share/man/es/man1/which.1.gz which.es1.gz /usr/share/man/es/man1/which.debianutils.1.gz --slave /usr/share/man/fr/man1/which.1.gz which.fr1.gz /usr/share/man/fr/man1/which.debianutils.1.gz --slave /usr/share/man/it/man1/which.1.gz which.it1.gz /usr/share/man/it/man1/which.debianutils.1.gz --slav

In [ ]:
%%bash

cd
pwd

# Create a test directory and navigate into it
mkdir -p test_dir/level1
cd test_dir
pwd

# Create some files and subdirectories
touch file1 file2 file3
mkdir subdir1 subdir2
ls # Default listing of files and directories in the current directory

# List in long format (-l) showing permissions, owner, group, size, and timestamp
ls -l

# List with human-readable file sizes (-lh) to display sizes in KB, MB, etc.
ls -lh

# Show hidden files as well (-a), includes files starting with '.'
ls -a

# Long format with hidden files (-la), useful to see permissions of hidden files
ls -la

# List in long format with hidden files and include inode number (-lia)
ls -lia

# Show the inode numbers (-i) of files and directories in the current directory
ls -i

# List in long format of only directories, showing details of directories themselves (-ld)
ls -ld */

# Show inode numbers along with hidden files in a directory listing (-ia)
ls -ia

# Show inode numbers along with detailed view for hidden files (-lia)
ls -lia

# Show file sizes in blocks (-s)
ls -s

# Recursive listing, show contents of all directories and subdirectories (-R)
ls -R

# Recursive listing in long format (-lR)
ls -lR

# Show files sorted by size (-lS) to get largest files first
ls -lS

# List files sorted by modification time, newest first (-lt)
ls -lt

# List files in reverse order by modification time (-ltr)
ls -ltr

# Combine -i and -l to show inode numbers along with detailed view (-il)
ls -il

# Combine -i, -a, and -l to show inode numbers, hidden files, and detailed view (-ial)
ls -ial

# Using a directory as an argument, list contents of "level1" directory
ls level1

# Navigate into the "level1" directory and list contents
cd level1
pwd
ls -l

# Go back to parent directory and list contents recursively with details (-lR)
cd ..
ls -lR

# Use -d to show information only about the directory itself, not its contents
ls -ld level1

# List all files including hidden ones and inodes, but just at the top-level directory (-iAd)
ls -iAd *

# List files with detailed view and human-readable sizes, sort by size (-lhS)
ls -lhS

# List directories themselves in detailed view with inode numbers (-ild)
ls -ild */

# Clean up: remove the created test directory
cd ..
rm -r test_dir


This sequence includes a broad range of options for ls:

-l: Long format.
-h: Human-readable sizes.
-a: Include hidden files.
-i: Show inode numbers.
-d: Show directory information only.
-R: Recursive listing.
-S: Sort by size.
-t: Sort by modification time.
-r: Reverse the order.
-ia: Include inode numbers and hidden files.

In [14]:
%%bash

cd
pwd

# Create a test directory structure with nested subdirectories and files
mkdir -p test_dir/level1/level2
cd test_dir
touch file1 file2 level1/file3 level1/level2/file4
mkdir subdir1 subdir2

# Default listing of the current directory
ls

# List contents of a specific directory using it as an argument
ls level1

# List contents of multiple directories (subdir1, level1) using directory names as arguments
ls subdir1 level1

# List contents of multiple directories in long format (-l)
ls -l subdir1 level1

# List the contents of level1 recursively (-R), showing all files in subdirectories
ls -R level1

# List detailed view of the level1 directory itself without showing its contents (-ld)
ls -ld level1

# Using -ld to show information about the directory "level1/level2" without listing its contents
ls -ld level1/level2

# List with detailed view and human-readable sizes (-lh) for multiple directories
ls -lh level1 subdir1

# Recursive listing for multiple directories (-lR)
ls -lR level1 subdir2

# Combine options to show all directories and subdirectories recursively with human-readable sizes (-alhR)
ls -alhR level1 subdir1 subdir2

# Use a combination of -i (inode numbers) and directory arguments
ls -i level1 level1/level2

# List inode numbers and long format with human-readable sizes for multiple directories (-ilh)
ls -ilh level1 subdir1 subdir2

# Clean up: remove the created test directory
cd ..
rm -r test_dir


/home/kavis
file1
file2
level1
subdir1
subdir2
file3
level2
level1:
file3
level2

subdir1:
level1:
total 4
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:47 file3
drwxr-xr-x 2 kavis kavis 4096 Oct  8 02:47 level2

subdir1:
total 0
level1:
file3
level2

level1/level2:
file4
drwxr-xr-x 3 kavis kavis 4096 Oct  8 02:47 level1
drwxr-xr-x 2 kavis kavis 4096 Oct  8 02:47 level1/level2
level1:
total 4.0K
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:47 file3
drwxr-xr-x 2 kavis kavis 4.0K Oct  8 02:47 level2

subdir1:
total 0
level1:
total 4
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:47 file3
drwxr-xr-x 2 kavis kavis 4096 Oct  8 02:47 level2

level1/level2:
total 0
-rw-r--r-- 1 kavis kavis 0 Oct  8 02:47 file4

subdir2:
total 0
level1:
total 12K
drwxr-xr-x 3 kavis kavis 4.0K Oct  8 02:47 .
drwxr-xr-x 5 kavis kavis 4.0K Oct  8 02:47 ..
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:47 file3
drwxr-xr-x 2 kavis kavis 4.0K Oct  8 02:47 level2

level1/level2:
total 8.0K
drwxr-xr-x 2 kavis kavis 4.0K Oct  8 02:47 .
drwxr-x

Explanation of ls with Directories as Arguments:
ls <directory_name>: Lists the contents of the specified directory, not the current working directory.

Example: ls level1 lists files inside level1 directory.
ls <dir1> <dir2>: Lists contents of multiple directories.

Example: ls level1 subdir1 lists files in both level1 and subdir1.
ls -ld <directory>: Lists information about the directory itself, not its contents.

Example: ls -ld level1 shows details (like permissions, owner, etc.) of the directory level1 itself.
Recursive Listing (ls -R <directory>): Displays all files and subdirectories within the specified directory recursively.

Example: ls -R level1 shows all files in level1 and its subdirectories.
Combine Options with Directory Arguments:

ls -alhR level1 subdir1: Shows all hidden files in a long format, human-readable sizes, and recursively lists contents of level1 and subdir1.

In [ ]:
%%bash

cd
pwd


# Create a test directory structure with nested subdirectories and files
mkdir -p test_dir/level1/level2
cd test_dir
touch file1 file2 level1/file3 level1/level2/file4
mkdir subdir1 subdir2

# 1. Default `ls` with no options
ls

# 2. Using `-l -t` vs `-t -l`:
#    `-l` is for long format and `-t` is for sorting by modification time.
#    These two options can be used in any order and the output will be the same.
ls -lt  # Long format sorted by modification time
ls -tl  # Sorted by modification time with long format

# 3. Using `-l` and `-S` in different orders:
#    `-S` sorts by size and `-l` shows long format. The order does not matter here.
ls -lS  # Long format sorted by file size
ls -Sl  # Sorted by file size in long format

# 4. Combining options `-l`, `-a`, and `-t` in different orders:
#    All options will work in any order, and the output will be the same.
ls -lat  # Long format, including hidden files, sorted by modification time
ls -tal  # Sorted by modification time, long format, including hidden files
ls -alt  # Including hidden files, long format, sorted by modification time

# 5. Using `-d` with directory arguments:
#    `-d` changes how directories are listed. The order of `-d` matters if you include it with other options.
#    When `-d` is specified after other options, it will change how the directories themselves are displayed.
ls -ld level1 subdir1  # Show details of directories level1 and subdir1, not their contents
ls -dl level1 subdir1  # Same output as above since -d modifies the behavior of -l

# 6. Recursive listing (`-R`) combined with other options in different orders:
#    `-R` will recursively list all contents. Combining with other options like `-l`, `-a`, or `-S` can be in any order.
ls -lRa  # Long format, recursive, including hidden files
ls -aRl  # Including hidden files, recursive, long format

# Clean up: remove the created test directory
cd ..
rm -r test_dir


Key Takeaways on Option Ordering:
Most options can be combined in any order:

For example, ls -lS, ls -Sl, and ls -l -S will produce the same output because -l and -S do not conflict.
Some options modify the behavior of other options:

ls -d (show directory itself) when combined with -l (ls -ld) will show details of the directory itself rather than its contents. But using -l -d or -d -l does not change the behavior.
Grouping of options:

Options can be grouped together as ls -l -t or ls -lt. There is no difference in output when options are grouped or not, as long as the options do not modify each other’s behavior.
Effect of -R (recursive):

The order of -R does not matter when combined with other options like -l or -a. All contents will be listed recursively no matter where -R appears.
Options that change display behavior (e.g., -d, -l):

These options need to be understood carefully, as -d will override the behavior of other options like -R (no recursion).
Summary
The order of options does not matter in most cases, but options like -d can change how others behave, so be mindful when combining.

In [15]:
%%bash

cd 
pwd

# Create a test directory structure for demonstration
mkdir -p test_recursive/dir1/dir2
cd test_recursive
touch file1 file2 dir1/file3 dir1/dir2/file4
mkdir dir3

# 1. Basic Recursive Listing (-R)
#    Lists the contents of the current directory and all its subdirectories.
ls -R

# 2. Recursive Listing in Long Format (-lR)
#    Long format listing, showing permissions, owners, sizes, timestamps, and file names recursively.
ls -lR

# 3. Recursive Listing with Hidden Files (-aR)
#    Includes hidden files (those starting with a dot `.`) in the listing.
ls -aR

# 4. Recursive Listing with Long Format and Hidden Files (-laR)
#    Combines long format and hidden files with recursive listing.
ls -laR

# 5. Recursive Listing for a Specific Directory (dir1)
#    Lists all contents of `dir1` and its subdirectories, recursively.
ls -R dir1

# 6. Recursive Listing for Multiple Directories (dir1, dir3)
#    Lists all contents of `dir1` and `dir3` and their subdirectories recursively.
ls -R dir1 dir3

# 7. Recursive Listing with Human-Readable Sizes (-lhR)
#    Shows sizes in KB, MB, etc., for better readability.
ls -lhR

# 8. Recursive Listing with Sorting by Modification Time (-ltR)
#    Recursively lists all contents, sorting each directory’s contents by modification time.
ls -ltR

# 9. Recursive Listing in Reverse Order (-lrR)
#    Recursively lists all contents, displaying each directory’s contents in reverse order.
ls -lrR

# 10. Recursive Listing and Limiting Output to Directories (-dR)
#     Show information about directories themselves, without listing their contents.
#     NOTE: -d overrides the recursive behavior, so only directories are listed.
ls -dR */

# 11. Combining -R with -s to Show File Sizes in Blocks
#     Displays file sizes in blocks recursively through the directory tree.
ls -sR

# Clean up: remove the created test directory
cd ..
rm -r test_recursive


/home/kavis
.:
dir1
dir3
file1
file2

./dir1:
dir2
file3

./dir1/dir2:
file4

./dir3:
.:
total 8
drwxr-xr-x 3 kavis kavis 4096 Oct  8 02:48 dir1
drwxr-xr-x 2 kavis kavis 4096 Oct  8 02:48 dir3
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:48 file1
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:48 file2

./dir1:
total 4
drwxr-xr-x 2 kavis kavis 4096 Oct  8 02:48 dir2
-rw-r--r-- 1 kavis kavis    0 Oct  8 02:48 file3

./dir1/dir2:
total 0
-rw-r--r-- 1 kavis kavis 0 Oct  8 02:48 file4

./dir3:
total 0
.:
.
..
dir1
dir3
file1
file2

./dir1:
.
..
dir2
file3

./dir1/dir2:
.
..
file4

./dir3:
.
..
.:
total 16
drwxr-xr-x  4 kavis kavis 4096 Oct  8 02:48 .
drwxr-x--- 11 kavis kavis 4096 Oct  8 02:48 ..
drwxr-xr-x  3 kavis kavis 4096 Oct  8 02:48 dir1
drwxr-xr-x  2 kavis kavis 4096 Oct  8 02:48 dir3
-rw-r--r--  1 kavis kavis    0 Oct  8 02:48 file1
-rw-r--r--  1 kavis kavis    0 Oct  8 02:48 file2

./dir1:
total 12
drwxr-xr-x 3 kavis kavis 4096 Oct  8 02:48 .
drwxr-xr-x 4 kavis kavis 4096 Oct  8 02:48 ..
drwx

Explanation of Recursive Listing Options:
ls -R: This is the basic recursive listing. It displays all files and directories under the current directory and its subdirectories. It’s useful for getting a complete overview of the entire directory tree.

ls -lR: Combines the long format listing (-l) with recursive listing. This provides a detailed view of permissions, owners, sizes, and modification timestamps for all files and directories recursively.

ls -aR: Shows all files, including hidden ones (files starting with .), recursively.

ls -laR: Combines long format and hidden file listing with recursive listing.

ls -R <directory_name>: Lists the contents of a specific directory (e.g., dir1) recursively. It will show all files and subdirectories under dir1.

ls -lhR: Combines long format with human-readable file sizes (KB, MB, etc.) for all files and directories recursively.

ls -ltR: Combines sorting by modification time (-t) with recursive listing. This will show the most recently modified files first within each directory.

ls -lrR: Combines reverse sorting (-r) with recursive listing. This will display files in reverse alphabetical order within each directory.

ls -dR: Using -d with -R will only display the directory names themselves, not their contents. This option can be useful if you want to list the directories without listing everything inside.

ls -sR: Shows file sizes in blocks, recursively. This is useful for getting an idea of storage consumption in each directory.

When to Use Recursive Listing:
To see the structure of nested directories.
To analyze large directory trees for storage usage or permissions.
To find specific files deep within a directory structure without manually traversing each directory.
Using recursive listing helps in scenarios where you want to understand the overall structure and contents of a directory tree.

In [ ]:
%%bash

cd
pwd

# Create a sample directory structure
mkdir -p sample_dir/subdir
cd sample_dir
touch file1 file2 .hidden_file subdir/file3

# Short form of options:
ls -a          # Show all files, including hidden files
ls -l          # Show long format listing
ls -la         # Combine -l and -a to show hidden files in long format
ls -lh         # Long format with human-readable file sizes
ls -R          # Recursively list all directories and their contents
ls -d */       # Show directory names only
ls -i          # Show inode numbers
ls -r          # Reverse the order of the listing
ls -S          # Sort by file size
ls -t          # Sort by modification time

# Long form of options:
ls --all           # Equivalent to -a, show all files including hidden ones
ls --long          # Equivalent to -l, show long format listing
ls --all --long    # Equivalent to -la, show hidden files in long format
ls --human-readable --long  # Long format with human-readable sizes
ls --recursive     # Equivalent to -R, show directories and their contents recursively
ls --directory */  # Equivalent to -d, show directory names only
ls --inode         # Equivalent to -i, show inode numbers
ls --reverse       # Equivalent to -r, show files in reverse order
ls --sort=size     # Equivalent to -S, sort by file size
ls --time          # Equivalent to -t, sort by modification time

# Clean up: remove the created test directory
cd ..
rm -r sample_dir


Key Points to Remember:
Short Form: Typically a single character preceded by a hyphen (-), e.g., -l, -a, -R.
Long Form: A descriptive word preceded by two hyphens (--), e.g., --all, --recursive, --long.
Short and long forms can be used interchangeably. For example, -l is equivalent to --long.
Multiple short options can be combined (e.g., ls -laR), but long options must be written separately (e.g., ls --long --all --recursive).
Use man ls or ls --help to see all available short and long forms for ls.

Short Form vs. Long Form of ls Options
Short Form	Long Form	Description
-a	          --all	    Show all files, including hidden files (those starting with .).
-A	       --almost-all	Show all files except . (current directory) and .. (parent).
-d	     --directory	List directory names only, not their contents.
-h	--human-readable	Show sizes in human-readable format (KB, MB, etc.).
-l	          --long	Use a long listing format to show detailed information.
-R	     --recursive	List subdirectories recursively.
-S	     --sort=size	Sort files by size (largest first).
-t	          --time	Sort by modification time (newest first).
-r	--reverse	Reverse the order of sorting.
-i	--inode	Show inode numbers.
-s	--size	Display file sizes in blocks.
-u	--access-time	Sort by last access time instead of modification time.
-c	--time=ctime	Sort by change time (status change time).
-G	--no-group	Do not show group information in the long format.
-X	--sort=extension	Sort alphabetically by file extension.
-1	--format=single-column	List one file per line.

In [1]:
%%bash
cd /
cd var/log
pwd 
chmod o+w alternatives.log  # permissions change not allowed by users other than root directory 
mkdir level1                # write not allowed by users other than root directory 
touch file1
clear

/var/log


chmod: changing permissions of 'alternatives.log': Operation not permitted
mkdir: cannot create directory ‘level1’: Permission denied
touch: cannot touch 'file1': Permission denied


In [18]:
%%bash
pwd
cd 
pwd
cd rep1/Linux_shell
file Linux_shell_w1.ipynb   #what kind of file is it
file 1500x500.jpeg          #its a binary file, not a text file
file random_toss.py
file level1

/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
/home/kavis
Linux_shell_w1.ipynb: Unicode text, UTF-8 text, with very long lines (1311)
1500x500.jpeg: JPEG image data, JFIF standard 1.01, aspect ratio, density 1x1, segment length 16, progressive, precision 8, 1500x500, components 3
random_toss.py: ASCII text, with CRLF line terminators
level1: directory


In [22]:
%%bash
date

cd /rep1/Linux_shell
ls -l

touch SC_Week0_Unit0.ipynb   #changes the timestamp to latest time
ls -l


Sun Oct  6 19:54:49 UTC 2024


bash: line 3: cd: /rep1/Linux_shell: No such file or directory


total 544
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis  90847 Oct  6 19:32 Linux_shell_w1.ipynb
-rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis  16186 Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavis kavis   3531 Oct  5 14:22 SC_Week0_Unit4.ipynb
-rwxrwxrwx 1 kavis kavis   4159 Oct  5 14:21 SC_Week0_Unit5.ipynb
drwxrwxrwx 1 kavis kavis   4096 Oct  6 18:19 level1
-rwxrwxrwx 1 kavis kavis    199 Oct  6 19:47 random_toss.py
total 544
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis  90847 Oct  6 19:32 Linux_shell_w1.ipynb
-rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-r

In [10]:

%%bash 

cd 
pwd

cd ..
pwd

cd ..
pwd

cd ..
pwd

cd //home///kavis
pwd


/home/kavis
/home
/
/
//home/kavis


Multiple uses of / is as good as one
The root folder / is its own parent 